packages/zoltan2/src/algorithms/Zoltan2_Algorithm.hpp

```diff
@@ -115,6 +116,91 @@ class Algorithm {
     Z2_THROW_NOT_IMPLEMENTED
   }
 
+  //! \brief  for partitioning methods, return partition tree of the
+  //          computed parts
+  //
+  virtual const std::vector<partitionTreeNode<part_t> > &
+  getPartitionTreeNodes() const
+  {
+    return partitionTree;
+  }
+
+  //! \brief  for partitioning methods, set partition tree of the
+  //          computed parts.
+  //
+  void setPartitionTreeNodes(
+    const std::vector<partitionTreeNode<part_t> > & setPartitionTree)
+  {
+    partitionTree = setPartitionTree;
+
+    // now by convention the last node is the root
+    // this would not be the natural result of rcb
+    // however other algorithms may also need this swap step so
+    // I made it general here
+    part_t findRoot = -1;
+    for(part_t n = 0; n < static_cast<part_t>(partitionTree.size()); ++n) {
+      if(partitionTree[n].parent == 0) { // root has parent=0 by convention
+        if(findRoot != -1) {
+          throw std::logic_error("setPartitionTreeNodes found more than one root. "
+            "This is not expected.");
+        }
+        findRoot = n;
+      }
+    }
+
+    if(findRoot == -1) {
+      throw std::logic_error( "setPartitionTreeNodes did not find a root. "
+        "This is not expected." );
+    }
+
+    // potentially we would not do this and simply apply the 'swap' on the
+    // data write out when requested - but that could create confusion
+    // To do - verify how we want to apply this convention
+    swapPartitionTreeNodes(findRoot, static_cast<part_t>(partitionTree.size())-1);
+
+    // Swap the order of the two nodes under the root
+    // This would intentionally put permPartNums out of sequence for rcb for ex.
+    // Just for validating code right now...
+    /*
+    part_t newRoot = partitionTree.size()-1;
+    part_t save1 = partitionTree[newRoot].children[1];
+    partitionTree[newRoot].children[1] = partitionTree[newRoot].children[0];
+    partitionTree[newRoot].children[0] = save1;
+    */
+  }
+
+  //! \brief  swap to node indices and update all the corresponding parent and
+  //          children indices to preserve the tree structure.
+  //
+  void swapPartitionTreeNodes(part_t a, part_t b) {
```

Comment:
I am surprised that the implementation of these methods is in the base Algorithm class. Other similar methods (e.g., boxAssign) are implemented in the specific algorithms in ways that are appropriate to the specific algorithms; PartitioningSolution then provides a call-through to the algorithm-specific implementation.  Indeed, some of the comments added to this file indicate one might do these operations differently for RCB than for other partitioners.